clang/lib/AST/StmtProfile.cpp

```diff
@@ -328,20 +328,20 @@ void StmtProfiler::VisitGCCAsmStmt(const GCCAsmStmt *S) {
   VisitStmt(S);
   ID.AddBoolean(S->isVolatile());
   ID.AddBoolean(S->isSimple());
-  VisitStringLiteral(S->getAsmString());
+  VisitExpr(S->getAsmStringExpr());
```

Comment:
Are these necessary? I would expect the `children` definition of the statement should instead print these.  It'll change how the ast-dump works, but for the better IMO.  Though perhaps we like how these look...

Reply:
I was able to revert!

Reply:
I would expect the string-literal to be an expression stored as well, right? So it should print with the children in the AST.

Reply:
it's always an expression (which sometimes is a string literal). but there are never both a string literal and an other expression

Reply:
We would need a specific iterator type as we are not iterating over a single sequence of elements.
There is a pre-existing defect that we don't use trailing objects for the children, which is something we might want to do too, but it seems unrelated to this change.

Also, profiling can certainly done in an arbitrary order but ideally if we were to fix that we would interleave constraints and their corresponding expressions so that ASTDump
makes more sense


- template string
- input constraint, input constraint expression, etc...
- output constraint, output constraint expression, etc...
- clobbers
- label

Either way, I would rather keep the change separate

Reply:
#131358